pandas/_libs/tslibs/nattype.pxd

```diff
@@ -1,6 +1,10 @@
 # -*- coding: utf-8 -*-
 # cython: profile=False
 
+from numpy cimport int64_t
+cdef int64_t NPY_NAT
+
 cdef bint _nat_scalar_rules[6]
 
-cdef bint _checknull_with_nat(object val)
+cdef bint checknull_with_nat(object val)
+cdef bint check_all_nulls(object val)
```

Comment:
where is definition for check_all_nulls? it shouldn't be here anyhow, that's a very general routine and should be in an inference module (I know we have a problem with Period, so can't create a inference.pyx really, though could come pretty close)

Reply:
Status quo it is in tslib.  In this PR it is in tslibs.nattype.  I'd be fine putting it back in src/inference.pyx.  The main thing I have in mind here is avoiding/reducing the dependence of `lib` on `tslib`.

> (I know we have a problem with Period, so can't create a inference.pyx really, though could come pretty close)

I don't follow.  Can you expand on this a little?

Reply:
I am all for reducing lib dependence on tslib. though I don't think you can eliminate it, after all Timestamp is defined there and Period in period.pyx

However inference.pyx itself *should* become a separate module (depending on Timestamp, Period, natttype).

I think then we could pull out the checknull* routines and simply put them in inference and it would work. lib wouldn't need it (or maybe it would but it would be just another module).

Reply:
> I am all for reducing lib dependence on tslib. though I don't think you can eliminate it, after all Timestamp is defined there and Period in period.pyx

`lib` does import `Timestamp` but never uses it.  And I don't think it imports `Period`.

> However inference.pyx itself should become a separate module

+1

I'd advocate mirroring the layout of `core`, so `_libs.dtypes.inference`.

Reply:
yep that would work.

right I think lib could prob become independent of tslib entirely (but might for example need inference) and prob nattype.